### PR TITLE
Add expression bodied function support

### LINKED ISSUE #20686: Add Expression-bodied functions
Lang Issue to track https://github.com/ballerina-platform/ballerina-spec/issues/395

### LINKED ISSUE #20990: Tooling support for expression bodied functions
PR #20929 adds support for expression bodied functions. Along with that, it introduces some changes to the way functions are modeled. Need to fix the tooling side for this.

### PR DESCRIPTION
**Pull request by pubudu91**
```ballerina
function toEmployee(Person p) returns Employee => {
    name: p.fname + " " + p.lname + " " + x.toString()
};

public function main() {
    Person p = {fname: "John", lname: "Doe", age: 25};
    io:println(toEmployee(p));
}
```

Fixes #20686
Fixes #20990

```ballerina
{ return expr; }
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/ExternFunctionBodyNode.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org.ballerinalang.model.tree;
+
+/**
+ * Represents the body of a function whose implementation is native.
+ *
+ * @since 1.2.0
+ */
+public interface ExternFunctionBodyNode extends FunctionBodyNode {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/ExternFunctionBodyNode.java]: Change to ExternalFunctionBodyNode

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java [hunk 2]
```diff
@@ -283,7 +283,44 @@ public void exitBlockFunctionBody(BallerinaParser.BlockFunctionBodyContext ctx)
...
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void exitExprFunctionBody(BallerinaParser.ExprFunctionBodyContext ctx) {
+        if (isInErrorState) {
+            return;
+        }
+
+        // TODO: Check if we need to take pos and ws of expr or the whole body
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: Remove the todo

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 3]
```diff
@@ -1926,6 +1960,9 @@ void endExternalFunctionBody(int annotCount) {
         // reversing the collected annotations to preserve the original order
         Collections.reverse(tempAnnotAttachments);
         tempAnnotAttachments.forEach(invokableNode::addExternalAnnotationAttachment);
+
+        FunctionBodyNode body = this.funcBodyNodeStack.pop();
+        invokableNode.setBody(body);
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: move the extern function related logic to the extern func body from the invoke node.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -316,14 +319,15 @@ public void visit(BLangFunction funcNode) {
 
         // Check for native functions
         if (Symbols.isNative(funcNode.symbol) || funcNode.interfaceFunction) {
-            if (funcNode.body != null) {
+            // TODO: double check this. Only applies for interfaces now IINM.
+            if (funcNode.funcBody != null) {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Change this along with the extern body changes

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 5]
```diff
@@ -390,6 +393,29 @@ public void visit(BLangFunction funcNode) {
...
+                break;
+            } else {
+                analyzeNode(stmt, bodyEnv);
+            }
+        }
+    }
+
+    @Override
+    public void visit(BLangExprFunctionBody body) {
+        SymbolEnv bodyEnv = SymbolEnv.createFuncBodyEnv(body, env);
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Check and see whether we can reuse the envs created in prev stages

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangFunctionBody.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BVarSymbol;
+
+/**
+ * Represents a function/method body.
+ *
+ * @since 1.2.0
+ */
+public abstract class BLangFunctionBody extends BLangNode implements FunctionBodyNode {
+
+    public BVarSymbol mapSymbol; // TODO: Do we need closure map for non-block bodied functions?
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangFunctionBody.java]: Check and remove this

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/HttpFiltersDesugar.java [hunk 7]
```diff
@@ -233,7 +234,8 @@ private BLangSimpleVariable addFilterContextCreation(BLangFunction resourceNode,
                 new BVarSymbol(0, names.fromString(filterContextVarName), resourceNode.symbol.pkgID, filterContextType,
                                resourceNode.symbol));
         filterContextVar.typeNode = filterContextUserDefinedType;
-        resourceNode.body.stmts.add(0, ASTBuilderUtil.createVariableDef(resourceNode.pos, filterContextVar));
+        ((BLangBlockFunctionBody) resourceNode.funcBody).stmts.add(0, ASTBuilderUtil
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/HttpFiltersDesugar.java]: Consider other types of bodies as well

**Comment by pubudu91** [comment]
Add tests for taint checking

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 8]
```diff
@@ -372,23 +404,63 @@ public void visit(BLangSimpleVariableDef varDefNode) {
...
+            return createMapSymbolIfAbsent((BLangBlockFunctionBody) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.BLOCK) {
+            return createMapSymbolIfAbsent((BLangBlockStmt) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.FUNCTION) {
+            return createMapSymbolIfAbsent((BLangFunction) node, closureMapCount);
+        }
+        return null;
+    }
+
+    private BVarSymbol createMapSymbolIfAbsent(BLangBlockFunctionBody body, int closureMapCount) {
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: Can ```getOrCreateMapSymbol``` be a better name.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 9]
```diff
@@ -372,23 +404,63 @@ public void visit(BLangSimpleVariableDef varDefNode) {
...
 
+    private BVarSymbol createMapSymbolIfAbsent(BLangNode node, int closureMapCount) {
+        if (node.getKind() == NodeKind.BLOCK_FUNCTION_BODY) {
+            return createMapSymbolIfAbsent((BLangBlockFunctionBody) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.BLOCK) {
+            return createMapSymbolIfAbsent((BLangBlockStmt) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.FUNCTION) {
+            return createMapSymbolIfAbsent((BLangFunction) node, closureMapCount);
+        }
+        return null;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: The name suggests that we create a map symbol always but the function may return null as well. Should this null be an assert ?

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: `null` return is for the cases where a closure var map symbol is not applicable.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java (hunk 8 above)
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: Sort of gives the idea that the method has multiple responsibilities doesn't it?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java (hunk 9 above)
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: To further clarify this, for nodes for which closures are not applicable, calling this method doesn't have any effect and we don't do anything with the return value. We use the return value returned by this method only in contexts where the closure map is applicable.

**Pull request merged**